code/graphics/generic.cpp

```diff
@@ -500,6 +583,115 @@ void generic_anim_render(generic_anim *ga, float frametime, int x, int y, bool m
 		else {
 			gr_set_bitmap(ga->first_frame + ga->current_frame);
 		}
+	}
+}
+
+/*
+ * @brief calculate current frame for variable frame delay animation formats (e.g. apng)
+ *
+ * @param [in] *ga  animation data
+ * @param [in] frametime  how long this frame took
+ */
+void generic_anim_render_variable_frame_delay(generic_anim* ga, float frametime)
+{
+	Assertion(ga->type == BM_TYPE_PNG, "only valid for apngs (currently); get a coder!");
+	if (ga->keyframe != 0) {
+		Warning(LOCATION, "apngs don't support keyframes");
+		return;
+	}
+
+	// don't change the frame time if we're paused
+	if((ga->direction & GENERIC_ANIM_DIRECTION_PAUSED) == 0) {
+		if(ga->direction & GENERIC_ANIM_DIRECTION_BACKWARDS) {
+			// playing backwards
+			ga->anim_time -= frametime;
+			if((ga->direction & GENERIC_ANIM_DIRECTION_NOLOOP) && ga->anim_time <= 0.0) {
+				ga->anim_time = 0;  //stop on first frame when playing in reverse
+			}
+			else {
+				if (ga->anim_time <= 0.0) {
+					// loop back to end
+					ga->anim_time = ga->total_time - 0.001f;
+					ga->png.previous_frame_time = ga->total_time;
+					ga->png.anim->current_frame = ga->num_frames-1;
+					ga->current_frame = ga->num_frames-1;
+				}
+			}
+		}
+		else {
+			// playing forwards
+			ga->anim_time += frametime;
+			if(ga->anim_time >= ga->total_time) {
+				if(ga->direction & GENERIC_ANIM_DIRECTION_NOLOOP) {
+					ga->anim_time = ga->total_time - 0.001f;  // stop on last frame when playing - if it's equal fmod sets back to 0.0f
+				}
+				else {
+					// loop back to start
+					ga->png.previous_frame_time = 0.0f;
+					ga->png.anim->current_frame = 0;
+					ga->current_frame = 0;
+				}
+				ga->done_playing = 1;
+			}
+		}
+		ga->anim_time = fmod(ga->anim_time, ga->total_time);
+	}
+
+	if (ga->num_frames > 0) {
+
+		// just increment or decrement the frame by one
+		// jumping forwards multiple frames will just cause slowdowns as multiple frames
+		// would need to be composed
+		if (ga->direction & GENERIC_ANIM_DIRECTION_BACKWARDS) {
+			if (ga->anim_time <= ga->png.previous_frame_time - ga->png.anim->frame.delay) {
+				ga->png.previous_frame_time -= ga->png.anim->frame.delay;
+				ga->current_frame--;
+			}
+		}
+		else {
+			if (ga->anim_time >= ga->png.previous_frame_time + ga->png.anim->frame.delay) {
+				ga->png.previous_frame_time += ga->png.anim->frame.delay;
+				ga->current_frame++;
+			}
+		}
+
+		// verbose debug; but quite useful
+		nprintf(("apng", "apng generic render timings/frames: %04f %04f %04f %04f | %03i %03i %03i\n",
+				frametime, ga->anim_time, ga->png.anim->frame.delay, ga->png.previous_frame_time,
+				ga->previous_frame, ga->current_frame, ga->png.anim->current_frame));
+
+		if(ga->streaming) {
+			generic_render_png_stream(ga);
+			gr_set_bitmap(ga->bitmap_id);
+		}
+		else {
+			Error(LOCATION, "non-streaming apngs not implemented yet");
```

Comment:
Is this an error that can be caused by user data or only if a programmer uses the wrong API? If it can only be caused by bad code then the condition could be removed and converted into an `Assertion`.